File: android/app/src/main/java/com/openburnbar/data/policy/MobileProviderAccountPolicy.kt
package com.openburnbar.data.policy

enum class MobileProviderConnectivity(val wire: String) {
    LOCAL_ONLY("local-only"),
    CLOUD_CONNECTED("cloud-connected"),
}

enum class MobileProviderErrorClass(val wire: String, val userVisibleLabel: String) {
    DENIED("denied", "Permission denied"),
    OFFLINE("offline", "Offline"),
    EXPIRED("expired", "Credential expired"),
    MALFORMED("malformed", "Provider data is malformed"),
}

object MobileProviderAccountPolicy {
    fun connectivity(storageScope: String): MobileProviderConnectivity = when (storageScope) {
        "cloud_refreshable", "server_private" -> MobileProviderConnectivity.CLOUD_CONNECTED
        else -> MobileProviderConnectivity.LOCAL_ONLY
    }

    fun isCloudConnected(storageScope: String): Boolean =
        connectivity(storageScope) == MobileProviderConnectivity.CLOUD_CONNECTED

    fun classifyError(code: String, message: String? = null): MobileProviderErrorClass {
        val haystack = listOf(code, message.orEmpty()).joinToString(" ").replace(" ", "").lowercase()
        return when {
            "permission-denied" in haystack || "permissiondenied" in haystack || "denied" in haystack ->
                MobileProviderErrorClass.DENIED
            "unavailable" in haystack || "network" in haystack || "offline" in haystack ->
                MobileProviderErrorClass.OFFLINE
            "expired" in haystack || "deadline-exceeded" in haystack ->
                MobileProviderErrorClass.EXPIRED
            else -> MobileProviderErrorClass.MALFORMED
        }
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarMobilePolicy/MobileProviderAccountPolicy.swift
import Foundation

public enum MobileProviderConnectivity: String, Sendable, Equatable {
    case localOnly = "local-only"
    case cloudConnected = "cloud-connected"
}

public enum MobileProviderErrorClass: String, Sendable, Equatable, CaseIterable {
    case denied
    case offline
    case expired
    case malformed

    public var userVisibleLabel: String {
        switch self {
        case .denied: return "Permission denied"
        case .offline: return "Offline"
        case .expired: return "Credential expired"
        case .malformed: return "Provider data is malformed"
        }
    }
}

public enum MobileProviderAccountPolicy {
    public static func connectivity(storageScope: String) -> MobileProviderConnectivity {
        switch storageScope {
        case "cloud_refreshable", "server_private":
            return .cloudConnected
        default:
            return .localOnly
        }
    }

    /// Local-only / device-keychain accounts never count as cloud-connected.
    public static func isCloudConnected(storageScope: String) -> Bool {
        connectivity(storageScope: storageScope) == .cloudConnected
    }

    public static func classifyError(code: String, message: String? = nil) -> MobileProviderErrorClass {
        let haystack = [code, message ?? ""]
            .joined(separator: " ")
            .replacingOccurrences(of: " ", with: "")
            .lowercased()
        if haystack.contains("permission-denied") || haystack.contains("permissiondenied")
            || haystack.contains("denied") {
            return .denied
        }
        if haystack.contains("unavailable") || haystack.contains("network") || haystack.contains("offline")
            || haystack.contains("deadline-exceeded") {
            return .offline
        }
        if haystack.contains("expired") && !haystack.contains("deadline-exceeded") {
            return .expired
        }
        return .malformed
    }
}
